arch/arc/core/arc_smp.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2019 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief codes required for smp support
+ *
+ */
+#include <device.h>
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <ksched.h>
+#include <soc.h>
+#include <init.h>
+
+
+#ifndef IRQ_ICI
+#define IRQ_ICI 19
+#endif
+
+#define ARCV2_ICI_IRQ_PRIORITY 1
+
+static void sched_ipi_handler(void *unused)
+{
+	ARG_UNUSED(unused);
+
+	z_arc_connect_ici_clear();
+	z_sched_ipi();
+}
+
+/**
+ * @brief Check whether need to do thread switch in isr context
+ *
+ * @details u64_t is used to let compiler use (r0, r1) as return register.
+ *  use register r0 and register r1 as return value, r0 has
+ *  new thread, r1 has old thread. If r0 == 0, it means no thread switch.
+ */
+u64_t z_arch_smp_switch_in_isr(void)
+{
+	u64_t ret = 0;
+	u32_t new_thread;
+	u32_t old_thread;
+
+	if (!_current_cpu->swap_ok) {
+		return 0;
+	}
+
+	old_thread = (u32_t)_current;
+
+	new_thread = (u32_t)z_get_next_ready_thread();
+
+	if (new_thread != old_thread) {
+		_current_cpu->swap_ok = 0;
+		((struct k_thread *)new_thread)->base.cpu =
+				z_arch_curr_cpu()->id;
+		_current = (struct k_thread *) new_thread;
+		ret = new_thread | ((u64_t)(old_thread) << 32);
+	}
+
+	return ret;
+}
+
+volatile struct {
+	void (*fn)(int, void*);
+	void *arg;
+} arc_cpu_init[CONFIG_MP_NUM_CPUS];
+
+
+volatile u32_t arc_cpu_wake_flag;
+volatile u32_t _curr_irq_stack[CONFIG_MP_NUM_CPUS];
```

Comment:
some comments would help here. Why is the variable name _curr_irq_stack, _curr_ stands for current? Isn't it the per-cpu interrupt stack, already part of per CPU kernel context?

Reply:
_curr_irq_stack is used to simplfy the record of per_cpu interrupt stack. Because  zephyr's high level design, 
per_cpu stack pointly is deep in cpu struct,  not easy to use in assembly

Reply:
ok, understood.
With 'comment', I meant comment in the code, not comment in the review, can you please add the explanation?

Reply:
comments added